Add instructions for plain Windows / Visual Studio Install

ralphlange opened the pull request:
Split detailed instructions in two parts.
Add a new top page for Windows with the common parts.

ralphlange committed:
Add a new Windows installation top page

ralphlange committed:
Split Windows installation in two parts (MSYS2 and plain)

ralphlange committed:
Thorough review and update of the MSYS2 installation document.

ralphlange committed:
Thorough review and update of the plain installation document.

anjohnson commented:
Hi Ralph,

I agree that this split makes more sense. Maybe the "Create a demo/test IOC" section could be split out of the MSYS part too and modified as necessary to work with both installation types?

I'm not keen on the "Python (if not already, very soon)" bullet-point on the main page, or on Python being included in the other pages. Python is not used at all for Windows builds yet (`$ORIGIN` isn't relevant on Windows at all) and I really don't see it replacing Perl in the current build system. Since Python isn't actually required, we shouldn't be telling people they need to install it to be able to build or use Base.

Would it be worth recommending the use of Chocolatey to install packages on the non-MSYS page? I find using that much easier than installing each package by hand, and we can also recommend using it for installing 7zip, git and vim if they are needed. wget isn't available, but curl is.

The following wording is self-contradictory: "The current directory mustn’t contain any spaces or parentheses. If it does, ...".

s/Start Manu/Start Menu/

We ship the files startup/win32.bat and startup/windows.bat which were intended for hand-editing to set up user environments on Windows, although they don't cover MSYS2. It would be nice to point people at them instead of having them edit their system environment variables. However as the Perl people say, there's more than one way to do it (whereas the Zen of Python, PEP 20 says "There should be one – and preferably only one – obvious way to do it").

Thanks for updating this, and for getting RTD PRs working too!

- Andrew

ralphlange commented:
Thanks for the review, Andrew!

I agree that the IOC-creating example should be a third section/page and have instructions that are generic enough to work for all. Remember, though, that MSYS / command prompt are different, MinGW / VS are different and 32bit / 64bit are different, so we're looking at eight slightly different sets of instructions. Not easy to fit into one readable page - which is exactly why I didn't do it yet.

I was putting the Python in just for you. ;-)
While I don't see a good reason not to install Python on a machine, I will take it out until it is strictly required (next year?).

I never trusted or used the startup batch scripts in Base in my life, that's why it didn't occur to me to point to them. I will shorten Hitesh's instructions of the system/user wide settings and point to the scripts as an alternative.
If you just have a single EPICS installation on the machine and want to run IOCs by clicking icons, the system settings might be more appropriate.

Chocolatey needs Admin rights, MSYS2 doesn't. I'm reluctant to recommend tools that I'm not able to use.
I will mention it, as it seems to be popular and working well.

HiteshDhola commented:
Hi Ralph,

I think it's a great Idea to have separate section for plain windows installation. I myself was working on a standard installation procedure using strawberry perl. I successfully installed it and was preparing the documentation. One obvious advantage against msys2 is that you can avoid longer path name. Your EPICS folder can be directly in c:/base, which is not possible in MSYS2 and path looks longer like this C:\Users\'user'\base. I will try to reflect this in new documentation.

Another suggestion:
In Plain windows installation, MSYS2 still is mentioned. It would look more clear if we avoid MSYS2 in plain windows section and do everything on command prompt.

Demo Ioc
I agree with Ralph, about splitting "demo Ioc" section. Procedure is more clear when mentioned separately for both rather than common method which might look confusing to users due to different instruction.

For environment variables,
Windows user generally doesn't use scripts (because they are windows users). They would be more comfortable using GUI for setting up Environment variables setup. But If we can make them use EPICS scripts with simple instructions in documentation, that would be great.

Hitesh

ralphlange commented:
The confusing situation (MSYS mentioned in the "plain" part) basically means that the simple separation MSYS vs. "plain" is not right.
There are four largely independent choices:
1. Which environment/shell to use for building
2. How to install the necessary tools
3. Which compiler to use for building
4. Which environment/shell to use for running

Any combination is possible. I will try to think about a better way to present this.

Python is not required yet. I will remove the mention from the How-To. (Personally, I expect it to be required for the next major version - 7.1 - some time next year.)

The plain "gcc" for MSYS should not be necessary. That's probably some kind of incomplete configuration or install. (Note that your instructions only installed the MinGW "-gcc" packages while I am always installing the complete "-toolchain" metapackage.) The toolchain package is definitely more than necessary, but - as far as I recall - the gcc package alone is not enough.

ralphlange committed:
Rewrite of the Windows top page

- remove mention of Python
- better describe choices and options
- mention and link to Chocolatey

ralphlange commented:
Main page rewritten - the others still have to follow.

ralphlange commented:
There's a discussion on a specific commit (b1c02c621dab7590704e54c795ee400f06295959) where Hitesh suggests to describe three common and simple options:

1. Using Visual studio (Plain Windows - Visual Studio)
2. Using MSYS2 - MingGW
3. Using Strawberry Perl (Plain Windows - MinGW)

I would add a fourth page with an "expert" (developer/maintainer) setup using a single MSYS2 install for both MinGW and VisualStudio, 32bit and 64bit.

ralphlange committed:
Improvements and fixes of the MSYS2 page

ralphlange committed:
Streamline and verify both MSYS2 and plain instructions

ralphlange committed:
Add separate page for setting the environment

ralphlange committed:
One sweep of corrections after the PR build on RTD

ralphlange commented:
I think this is ready to merge now.

anjohnson commented on getting-started/installation-windows-plain.rst:
```diff
@@ -0,0 +1,314 @@
+^^^^^^^^
+Install Strawberry Perl or ActivePerl using the Windows installers available on their download pages.
+
+Strawberry Perl contains a suitable version of GNU Make. Otherwise, you can download a Windows executable that Andrew provides at https://epics.anl.gov/download/tools/make-4.2.1-win64.zip. Unzip it into a location (path must not contain spaces or parentheses) and add it to the system environment, so that
```
The last sentence on this line is incomplete, maybe just replace `, so that` with `.`

anjohnson commented on getting-started/installation-windows-plain.rst:
```diff
@@ -0,0 +1,314 @@
+   **Note:** The complete path of the current directory mustn't contain any spaces or parentheses. If your working directory path does, you can do another cd into the same directory, replacing every path component containing spaces or parentheses with its Windows short path (that can be displayed with ``dir /x``).
+4. Set the EPICS host architecture EPICS_HOST_ARCH (windows-x86 for 64bit builds, win32-x86 for 32bit builds).
+5. Run the ``vcvarsall.bat`` script of your installation (the exact path depends on the type and language of installation) to set the environment for your build.
+6. Run ``make``.
```
or ``gmake`` if using the Strawberry Perl version.

anjohnson commented on getting-started/installation-windows-plain.rst:
```diff
@@ -0,0 +1,314 @@
+
+There will probably be warnings, but there should be no error. You can choose any EPICS base to install, the procedure remains the same.
+
+Quick test from MSYS2 Bash
```
This is the non-MSYS2 version, Bash isn't installed...

anjohnson reviewed (commented):
2 minor tweaks and one thing that might take a little longer.

HiteshDhola commented on getting-started/installation-windows-msys2.rst:
```diff
@@ -0,0 +1,290 @@
+    $ cd $HOME
+    $ wget https://epics-controls.org/download/base/base-7.0.4.1.tar.gz
+    $ tar -xvf base-7.0.4.1.tar.gz
+    $ cd base-R7.0.4.1
```
Can we rename EPICS base folder to  "base" ? So, that when new EPICS version is installed, one don't have to set Environment Variables again and restart the system.

In that case,
    $ cd $HOME
    $ wget https://epics-controls.org/download/base/base-7.0.4.1.tar.gz
    $ tar -xvf base-7.0.4.1.tar.gz
    $ mv base-R7.0.4.1 base
    $ cd base
    $ export EPICS_HOST_ARCH=windows-x64-mingw
    $ make

ralphlange replied:
You never have to restart the system after setting environment variables.

It is strongly suggested to keep base and support modules in versioned directories. This is how the whole build system is supposed to work: downstream modules switch between versions through their `configure/RELEASE` file. Otherwise, you will never be able to cleanly switch between dependency versions and rebuild modules, and for any given machine or module you will have a hard time finding out which version of base is installed and which version other modules were built against.

Examples should give an example (!) and not establish wrong procedures.

HiteshDhola commented on getting-started/installation-windows-msys2.rst:
```diff
@@ -0,0 +1,290 @@
+
+When using the MinGW 32bit toolchain, the "MSYS MinGW 32-bit" shell must be used and EPICS_HOST ARCH must be set to "win32-x86-mingw".
+
+During the compilation, there will probably be warnings, but there should be no error. You can choose any EPICS Base version to build, the procedure remains the same.
```
Can we specify build time ? like 30min to 60min depending on the system. When I first compiled EPICS I waited staring at screen for an hour expecting it to finished in 10-15 minutes.

ralphlange replied:
That depends on a lot of factors. Have a look at a recent AppVeyor build of EPICS 7 (https://ci.appveyor.com/project/epics-base-7/epics-base/builds/36440444) and you can see that on those machines, it's between 14 and 29 minutes, the 3.15 branch (https://ci.appveyor.com/project/epics-base-3/epics-base/builds/36371542) builds in 6 to 10 minutes.

These are relatively small two-core virtual machines. If you use a slow file system or a single core or a slow machine (or a combination), it can easily be an hour. On a faster machine (8 or 16 cores), it will be less.

ralphlange replied:
Point taken, though. Will add a sentence about build times.

HiteshDhola commented on getting-started/installation-windows-msys2.rst:
```diff
@@ -0,0 +1,290 @@
+
+Open a third shell. Using caput, modify the values of  ``test:pv1`` and ``test:pv2`` as we have done in the temperature example above. You will see changes of their sum in the second terminal accordingly.
+
+At this point, you have one IOC ``testioc`` running, which loaded the database ``test.db`` with 3 records. From other processes, you can connect to these records using Channel Access. If you add more process variable in ``test.db``, you will have to ``make`` the `testioc` application again and restart the IOC to load the new version of the database.
```
minor modification in last line :
 you will not have to ``make`` the `testioc` application again, just restart the IOC to load the new version of the database.

ralphlange replied:
You have to re-run make, else the database files (in `...App/Db`) will not get installed (in `TOP/db`) where the IOC loads them from.

HiteshDhola commented on getting-started/installation-windows-plain.rst:
```diff
@@ -0,0 +1,278 @@
+
+Open a third shell. Using caput, modify the values of  ``test:pv1`` and ``test:pv2`` as we have done in the temperature example above. You will see changes of their sum in the second terminal accordingly.
+
+At this point, you have one IOC ``testioc`` running, which loaded the database ``test.db`` with 3 records. From other processes, you can connect to these records using Channel Access. If you add more process variable in ``test.db``, you will have to ``make`` the `testioc` application again and restart the IOC to load the new version of the database.
```
minor modification in last line :
 you will not have to ``make`` the `testioc` application, just restart the IOC to load the new version of the database.

ralphlange replied:
You have to re-run make, else the database files (in `...App/Db`) will not get installed (in `TOP/db`) where the IOC loads them from.

HiteshDhola reviewed (approved):
That's all from me.

ralphlange committed:
Add section on build time

ralphlange committed:
Another sweep of updates

including contributions and feedback from Hitesh and Andrew
